### PR TITLE
Issue #381 #380: Add run_cfg_cmds and modify run_show_cmds

### PR DESCRIPTION
**Pull request by shachiagarwal**
# Please include a summary of the changes

vane/tests_tools.py - Modify run_show_cmds() to add timeout and new_conn option
vane/device_interface.py - Modify netmiko way of running config cmds

# Any specific logic/part of code you need extra attention on

To the logic of adding timer and new connection options. Please check if this is intitutive.

# Include the Issue number and link

#381 #380 

# List/Attach any dependencies/past issues that are required for this change/provide context

# Type of change

- - [ ] Bug fix (non-breaking change which fixes an issue)
- - [ ] New feature (non-breaking change which adds functionality)
- - [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- - [ ] This change requires a documentation update
- - [ ] Other (please specify)

# Effort required on reviewers end

- - [ ] Easy
- - [X] Medium
- - [ ] Hard 

# How Has This Been Tested?

Added tests in sample_network_tests
## Bug fix

    Include before and after snapshots/screenshots/changed logs
    
## New feature

    Ensure current test cases pass and include newer test cases if required
    
# CI pipeline result

- - [X] Pass
- - [ ] Fail
  
  If it fails, explain the reason and whether or not we should ignore the failure
  
# Verify Documentation Update

    If applicable, ensure documentation has been updated for ReadMe, Getting Started guide, and Style guide
    
# Additional comments

### CODE REVIEW

File: vane/tests_tools.py [hunk 1]
```diff
@@ -1213,23 +1240,141 @@ def set_evidence_default(self, dut_name):
...
+
+        Args: cfg_cmds: list of configuration cmds to run
+        dut: device on which cfg_cmds have to run
+        conn_type: connection type to dut - either pyeapi or netmiko
+        timeout: read timeout for dut connection
+        new_conn: whether to get a new conn to dut
+
+        Returns: A dict object that includes the response for each command
+        """
+        if dut is None:
```
[the remark this answers is not in the document]
Comment (shachiagarwal) [reply vane/tests_tools.py]: I have refactored the code to take this into account. Please review again. Thank you!

File: vane/tests_tools.py [hunk 2]
```diff
@@ -1255,48 +1372,79 @@ def run_show_cmds(self, show_cmds, dut=None, encoding="json", conn_type="eapi"):
...
+                int_cmd_txts.append(str(e))
+                # add the exception result to external evidence output list
+                if ext_cmd_txts is not None:
+                    ext_cmd_txts.append(e)
                 raise e
 
-            # add the show_clock_cmds to TestOps object's _show_cmds list
-            # also add the o/p of show_clock_cmds to TestOps object's _show_cmds_txts list
+            # add the show_clock_cmds to internal cmds list
+            # also add the o/p of show_clock_cmds to external cmd output list
```
Comment (rewati-arista) [inline vane/tests_tools.py:1404]: in the block you are adding output to internal cmd output but the comment says external cmd output, might want to correct it 😅

**Review by rewati-arista — APPROVED** [review]
LGTM apart from the minor rephrasing suggestions and one potential correction. 

Also @shachiagarwal since we have a lot of code refactored with the run_show_commands/run_cfg_command and run_record_cmds, do you think you could run us through the changed logic in these 3 methods in a review session so all of us are certain of the changed logic and on the same page with the understanding?

File: vane/tests_tools.py [hunk 3]
```diff
@@ -1052,6 +1057,26 @@ def _write_text_results(self):
             else:
                 logging.debug("No show command output to display")
 
+        # write evidence for cfg cmds if any
```
[the remark this answers is not in the document]
Comment (rewati-arista) [reply vane/tests_tools.py]: This looks good now!

File: vane/tests_tools.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rewati-arista) [reply vane/tests_tools.py]: Yep great job with the refactoring :)

File: sample_network_tests/vane_system_tests/test_vane.py [hunk 4]
```diff
@@ -292,3 +292,145 @@ def test_if_setup_fail_is_handled(self):
...
+
+            show_cmds = ["show running-config interfaces ethernet 16"]
+
+            tops.actual_output = tops.run_show_cmds(show_cmds, encoding="text")
+
+            tops.expected_output = f"description {description}"
+
+            cfg_cmds = ["interface eth16", f"no description {description}"]
+
+            tops.run_cfg_cmds(cfg_cmds, conn_type="ssh")
```
[the remark this answers is not in the document]
Comment (shachiagarwal) [reply sample_network_tests/vane_system_tests/test_vane.py:420]: Its testing both eapi and ssh and one test case. Just to demo, nothing changes by conn_type.

File: sample_network_tests/vane_system_tests/test_vane.py [hunk 5]
```diff
@@ -292,3 +292,145 @@ def test_if_setup_fail_is_handled(self):
...
+
+        except (AttributeError, LookupError, EapiError) as exception:
+            logging.error(
+                f"On device {tops.dut_name}: Error while running testcase on DUT is: "
+                f"{str(exception)}"
+            )
+            tops.actual_output = str(exception)
+            tops.output_msg += (
+                f"EXCEPTION encountered on device {tops.dut_name}, while "
+                f"investigating if ssh can be used to run cmds. Vane recorded error: {exception}"
```
[the remark this answers is not in the document]
Comment (shachiagarwal) [reply sample_network_tests/vane_system_tests/test_vane.py:430]: Correct

**Comment by shachiagarwal** [comment]
Please have a look @kgrozis-arista @rewati-arista @grybak-arista . Thanks!

**Comment by shachiagarwal** [comment]
Latest HTML report capturing this : 
Screenshot 2023-08-17 at 1 30 34 AM

File: tests/unittests/test_tests_tools.py [hunk 6]
```diff
@@ -1513,6 +1495,143 @@ def test_test_ops_run_show_cmds_text_exception_fail(mocker):
...
+    }
+
+    assert tops._show_cmd_txts == {
+        "DCBBW1": [
+            "Arista vEOS-lab\nHardware version: \nSerial number: SN-DCBBW1\n"
+            "Hardware MAC address: a486.49d7.e2d9\nSystem MAC address: a486.49d7."
+            "e2d9\n\nSoftware image version: 4.27.2F\nArchitecture: x86_64\n"
+            "Internal build version: 4.27.2F-26069621.4272F\nInternal build "
+            "ID: 2fd003fd-04c4-4b44-9c26-417e6ca42009\nImage format version"
+            ": 1.0\nImage optimization: None\n\nUptime: 3 days, 4 hours and "
```
Comment (grybak-arista) [inline tests/unittests/test_tests_tools.py:1555]: Is this essentially the content of OUTPUT (line 930) twice? Or the following:
```
"DCBBW1": [OUTPUT, OUTPUT]
```
Might be easier to read.

And the same for the block later on at line 1616.

**Review by grybak-arista — APPROVED** [review]
Looks good to me. 👍🏻 

Made a comment about some of the text in the unittests that may be able to be replaced with an existing variable, but not enough to prevent merging I think.

Approved!

File: tests/unittests/test_tests_tools.py (hunk 6 above)
Comment (shachiagarwal) [reply tests/unittests/test_tests_tools.py:1555]: :) will do that. Although I spent so many keyboard strokes to get it here with no warnings.

**Review by rewati-arista — APPROVED** [review]

File: tests/unittests/test_tests_tools.py (hunk 6 above)
Comment (rewati-arista) [reply tests/unittests/test_tests_tools.py:1555]: Have approved, @shachiagarwal once you get this change in we can merge!

**Pull request merged**